base/server/src/org/dogtagpki/server/rest/SecurityDomainHostService.java

```diff
@@ -0,0 +1,177 @@
+import java.util.Collection;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.ws.rs.core.GenericEntity;
+import javax.ws.rs.core.Response;
+
+import com.netscape.certsrv.base.BadRequestException;
+import com.netscape.certsrv.base.PKIException;
+import com.netscape.certsrv.base.ResourceNotFoundException;
+import com.netscape.certsrv.system.DomainInfo;
+import com.netscape.certsrv.system.SecurityDomainHost;
+import com.netscape.certsrv.system.SecurityDomainHostResource;
+import com.netscape.certsrv.system.SecurityDomainSubsystem;
+import com.netscape.cms.servlet.base.PKIService;
+import com.netscape.cms.servlet.csadmin.SecurityDomainProcessor;
+
+/**
+ * @author Endi S. Dewata
+ */
+public class SecurityDomainHostService extends PKIService implements SecurityDomainHostResource {
+
+    public static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(SecurityDomainHostService.class);
+
+    @Override
+    public Response getHosts() throws Exception {
+
+        logger.info("SecurityDomainService: Getting all security domain hosts");
+
+        Collection<SecurityDomainHost> hosts = new ArrayList<>();
+
+        SecurityDomainProcessor processor = new SecurityDomainProcessor(getLocale(headers));
+
+        DomainInfo domainInfo = processor.getDomainInfo();
+        logger.debug("SecurityDomainService: domain: " + domainInfo.getName());
+
+        for (SecurityDomainSubsystem subsystem : domainInfo.getSubsystems().values()) {
+            for (SecurityDomainHost host : subsystem.getHosts().values()) {
+                logger.debug("SecurityDomainService: - " + host.getId());
+                hosts.add(host);
+            }
+        }
+
+        GenericEntity<Collection<SecurityDomainHost>> entity =
+                new GenericEntity<Collection<SecurityDomainHost>>(hosts) {};
+
+        return createOKResponse(entity);
+    }
+
+    @Override
+    public Response getHost(String hostID) throws Exception {
+
+        logger.info("SecurityDomainService: Getting security domain host \"" + hostID + "\"");
+
+        SecurityDomainProcessor processor = new SecurityDomainProcessor(getLocale(headers));
+
+        DomainInfo domainInfo = processor.getDomainInfo();
+        logger.debug("SecurityDomainService: domain: " + domainInfo.getName());
+
+        for (SecurityDomainSubsystem subsystem : domainInfo.getSubsystems().values()) {
+            for (SecurityDomainHost host : subsystem.getHosts().values()) {
+
+                logger.debug("SecurityDomainService: - " + host.getId());
+
+                if (host.getId().equals(hostID)) {
+                    logger.debug("SecurityDomainService: Found security domain host " + hostID);
+                    return createOKResponse(host);
+                }
+            }
+        }
+
+        throw new ResourceNotFoundException("Security domain host not found: " + hostID);
+    }
+
+    @Override
+    public Response addHost(SecurityDomainHost host) throws Exception {
```

Comment:
Executing `securitydomain-host-add` twice with 2 different values replaces the value in LDAP with no warning. Is this desirable? 

Also, as mentioned in one of the above comment, the overwritten written values (eg: `adminPort`) are not displayed by any CLI

Reply:
It's not ideal, but that's how the current code works:
https://github.com/dogtagpki/pki/blob/master/base/server/src/com/netscape/cms/servlet/csadmin/SecurityDomainProcessor.java#L550

I think we should fix this behavior when we actually use the CLIs to replace the legacy code in pkispawn/pkidestroy, but for now I'd leave the code unchanged because it will require testing various installation scenarios.

Reply:
Agreed